### PR TITLE
Add signup endpoint

### LINKED ISSUE #6: [Feature] Create backend API to register users
**Is your feature request related to a problem? Please describe.**
There is no backend API to register users

**Describe the solution you'd like**
Create backend API to register users

POST | /api/v1/auth/signup | Creates user account
-- | -- | --

**Additional context**
#### Awesome :star::star::star::star::star:
```json
{
  "user": {
    "name": "john doe",
    "email": "john.doe@students.jkuat.ac.ke",
    "token": "jwt.token.here",
    "regno": "exx211-00xx/2020",
    "mobileno": "+2547xxxxxxxx"
  }
}
```
No authentication required, returns User data

Required fields: email, password, username, regno, mobileno

### PR DESCRIPTION
**Pull request by ObaraEmmanuel**
## Proposed changes

This pull requests adds the `api/v1/auth/signup` endpoint. This endpoint adds new users to the database and returns user data as well as JWT refresh and authentication tokens.

```json
{
    "access_token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ0b2tlbl90eXBlIjoiYW...",
    "refresh_token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ0b2tlbl90eXBlIjoice...",
    "user": {
        "name": "John Doe",
        "email": "hyper.space@students.jkuat.ac.ke",
        "regno": "ene212-00xx/20xx",
        "mobileno": "+254xxxxxxxxx"
    }
}
```
The struct

## Types of changes

What types of changes does your code introduce to votingBooth?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe):

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] Lint and unit tests pass locally with my changes
- [ ] Kept pull request small so they can be easily reviewed.
- [x] Make sure not to introduce any compiler warnings.
- [x] AVOID breaking the continuous integration build.
- [x] AVOID making significant changes to the driver's overall architecture.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### Where has this been tested?
This has not been tested yet

## Further comments

The returned payload is slightly different from the one required in #6. This is because I was using `dj-rest-auth` which generates such a payload when JWT authentication is enabled. Also, this PR is kinda huge because backend code was basically empty and I needed to write up a lot of boilerplate code to even be able to start tackling the issue at hand. Future PR's should be way shorter though.

### CODE REVIEW

**Review by rodneyosodo — CHANGES_REQUESTED** [review]
Check on the SECRET_KEY. The CI-CD tests aren't passing

**Comment by ObaraEmmanuel** [comment]
We will need to set the SECRET_KEY as an environment variable. I don't have access to the repository secrets or the workflow code so would you be able to do that?

**Comment by ObaraEmmanuel** [comment]
I have set up a place holder secret key ... not sure if the workflow changes on my fork will reflect.

File: .github/workflows/python-app.yml [hunk 1]
```diff
@@ -13,6 +13,8 @@ jobs:
   build:
 
     runs-on: ubuntu-latest
+    env:
```
Comment (rodneyosodo) [inline .github/workflows/python-app.yml]: @ObaraEmmanuel Remove this. I have added the SECRET_KEY to the GITHUB SECRETS

**Review by rodneyosodo — COMMENTED** [review]
This is fine

File: .github/workflows/python-app.yml (hunk 1 above)
Comment (ObaraEmmanuel) [reply .github/workflows/python-app.yml]: let me replace it with the github secret. The `env` has to be set either way

**Comment by rodneyosodo** [comment]
The tests are now passing. Great job. I think the changes are after the ci cd file was changed

**Comment by ObaraEmmanuel** [comment]
I have synchronized my branch with main now

**Comment by rodneyosodo** [comment]
@ObaraEmmanuel  Which file needs the env file?

**Comment by rodneyosodo** [comment]
> I have synchronized my branch with main now

Can you retry this again

**Comment by ObaraEmmanuel** [comment]
The .env is optional. As long as SECRET_KEY is set everything should work. I can't seem to be able to pull your new changes

File: src/software/backend/backend/settings/base.py [hunk 2]
```diff
@@ -1,29 +1,30 @@
...
+BASE_DIR = Path(__file__).resolve().parent.parent.parent
 
 
 # Quick-start development settings - unsuitable for production
-# See https://docs.djangoproject.com/en/3.2/howto/deployment/checklist/
+# See https://docs.djangoproject.com/en/3.1/howto/deployment/checklist/
 
 # SECURITY WARNING: keep the secret key used in production secret!
-SECRET_KEY = 'django-insecure-rf1mfkvesu-n0+kptvnw5ye-e%nqgs0s5&&j9%^5o690eui@^n'
+SECRET_KEY = os.getenv('SECRET_KEY')
```
Comment (rodneyosodo) [inline src/software/backend/backend/settings/base.py]: Reverts this

File: src/software/backend/backend/settings/base.py [hunk 3]
```diff
@@ -1,29 +1,30 @@
...
 
 # SECURITY WARNING: don't run with debug turned on in production!
-DEBUG = True
+DEBUG = False
```
Comment (rodneyosodo) [inline src/software/backend/backend/settings/base.py]: Reverts this

File: src/software/backend/backend/settings/base.py [hunk 4]
```diff
@@ -1,29 +1,30 @@
 """
-Django settings for backend project.
+Django settings for triple_edge_core project.
```
Comment (rodneyosodo) [inline src/software/backend/backend/settings/base.py]: Reverts this to `Django settings for backend project.`

File: src/software/backend/backend/settings/base.py [hunk 5]
```diff
@@ -1,29 +1,30 @@
...
 
-Generated by 'django-admin startproject' using Django 3.2.3.
+Generated by 'django-admin startproject' using Django 3.1.7.
```
Comment (rodneyosodo) [inline src/software/backend/backend/settings/base.py]: Revert this to `Generated by 'django-admin startproject' using Django 3.2.3.`

**Review by rodneyosodo — COMMENTED** [review]
Revert the indicated parts in the settings file

**Review by rodneyosodo — APPROVED** [review]

**Review by rodneyosodo — APPROVED** [review]

**Pull request merged**